### PR TITLE
Add unit tests pt 2

### PR DESCRIPTION
**Pull request by JeremyWesthead**
Should cover essentially all of the edge cases. Lots of minor fixes. Only rows missed by coverage are panics with contrived edge case inputs which would take longer to construct test cases for (basically constructing files for each edge case line)

### CODE REVIEW

**Comment by JeremyWesthead** [comment]
1300 lines of code, 5000 lines of tests :melting_face:

File: src/difference.rs [hunk 1]
```diff
@@ -162,33 +162,36 @@ impl GenomeDifference {
...
-                    let mut gene_name = None;
-                    let mut gene_position = None;
-                    let mut codon_idx = None;
 
                     if alt.alt_type == AltType::REF {
                         // Skip ref calls
                         continue;
                     }
 
                     // Only annotate with gene information if there is a single gene at this position
```
Comment (mcolpus) [inline src/difference.rs]: I think this comment is out of date now?

File: src/vcf.rs [hunk 2]
```diff
@@ -1173,18 +1178,35 @@ mod tests {
...
-                    ("COV".to_string(), vec!["0".to_string(), "68".to_string()]),
+                    ("RO".to_string(), vec!["66".to_string()]),
+                    ("AO".to_string(), vec!["2".to_string()]),
+                    ("COV".to_string(), vec!["66".to_string(), "2".to_string()]),
                     ("GT_CONF".to_string(), vec!["3.77".to_string()]),
                 ]),
                 is_filter_pass: false,
             },
+            VCFRow {
+                // Odd edge case which is a valid VCF row where it's a het GT but single COV value
```
Comment (mcolpus) [inline src/vcf.rs:1197]: How is that valid?? The FORMAT header for COV species a number as R meaning number of alleles including ref. As such it should have two values for COV. I think you are welcome to fail in this case should you wish

File: src/difference.rs [hunk 3]
```diff
@@ -1026,6 +1038,10 @@ impl GeneDifference {
 fn trim_float_string(mut float_string: String) -> String {
     while float_string.ends_with('0') {
         float_string.pop();
+        if float_string.ends_with("1.0") {
+            // Keep trailing 0 for 1.0
```
Comment (mcolpus) [inline src/difference.rs:1076]: do you want to comment on why?

File: src/genome.rs [hunk 4]
```diff
@@ -220,7 +220,7 @@ impl Genome {
         Assigns promoters to genes iteratively.
         Expand out to a given distance from the start of each gene without overlapping with other genes.
         */
-        let max_promoter_length = 100;
+        let max_promoter_length = 99;
```
Comment (mcolpus) [inline src/genome.rs:228]: Where is this number coming from

File: src/genome.rs [hunk 5]
```diff
@@ -4538,4 +5019,1648 @@ mod tests {
...
+                alt_nucleotides: None,
+                nucleotide_number: Some(-1),
+                nucleotide_index: Some(3),
+                indel_length: Some(-28),
+                indel_nucleotides: Some("aaaaaaaaccccccccccgggggggggg".to_string()),
+                amino_acid_number: None,
+                amino_acid_sequence: None,
+            },
+            Mutation {
+                mutation: "del_0.93".to_string(),
```
Comment (mcolpus) [inline src/genome.rs:5168]: I don't quite see where this comes from or what it means

File: src/genome.rs [hunk 6]
```diff
@@ -4538,4 +5019,1648 @@ mod tests {
...
+                        filter: vec!["PASS".to_string()],
+                        fields: HashMap::from([
+                            ("GT".to_string(), vec!["1/1".to_string()]),
+                            ("DP".to_string(), vec!["4".to_string()]),
+                            ("COV".to_string(), vec!["1".to_string(), "3".to_string()]),
+                            ("GT_CONF".to_string(), vec!["2.05".to_string()]),
+                        ]),
+                        is_filter_pass: true,
+                    },
+                    reference: "a".to_string(),
```
Comment (mcolpus) [inline src/genome.rs:5150]: What's the logic with the ref and alt swapping places compared to the vcf row

Comment (mcolpus) [reply src/genome.rs:5150]: Is alt just the bases removed for indels?

**Review by mcolpus — APPROVED** [review]

File: src/genome.rs (hunk 4 above)
Comment (JeremyWesthead) [reply src/genome.rs:228]: Matching `gumpy` behaviour. It maxes out at `-100` which is 99 bases rather than 100. It's arbitrary

File: src/genome.rs (hunk 5 above)
Comment (JeremyWesthead) [reply src/genome.rs:5168]: That's deletion of 93% of the gene - special representation for large deletions, documented in the `piezo` BNF for GARC

File: src/difference.rs (hunk 3 above)
Comment (JeremyWesthead) [reply src/difference.rs:1076]: It's just string formatting. Doesn't make a lot of sense to have `1.` rather than `1.0` (it's also not valid GARC without the trailing `0` in this case)

Comment (mcolpus) [reply src/difference.rs:1076]: oh, is this function only applied to numbers between 0 and 1? Otherwise 2.0 would have the same issue

Comment (JeremyWesthead) [reply src/difference.rs:1076]: Yeah, it's exclusively on divisions for percentages that it's called, so values should lie 0-1
It's basically a fun side effect of `format!({:.3})` giving trailing `0`s which aren't valid GARC (and are also just ugly)

File: src/vcf.rs (hunk 2 above)
Comment (JeremyWesthead) [reply src/vcf.rs:1197]: I'm not 100% sure either - I added the code for it due to a VCF I've seen, then contrived this test case to cover it
It's possible that it's defined as something other than `R` in some VCFs, but I try not to manually parse all of the ones I use
